src/Server.php

```diff
@@ -226,7 +238,24 @@ public function handleRequest(ConnectionInterface $conn, RequestInterface $reque
         );
 
         $callback = $this->callback;
-        $callback($request, $response);
+        $promise = new Promise(function ($resolve, $reject) use ($callback, $request) {
+            $resolve($callback($request));
+        });
+
+        $that = $this;
+        $promise->then(
+            function ($response) use ($that, $conn, $request) {
+                if (!$response instanceof ResponseInterface) {
+                    $that->emit('error', array(new \InvalidArgumentException('Invalid response type')));
+                    return $that->writeError($conn, 500);
+                }
+                $that->handleResponse($conn, $response, $request->getProtocolVersion());
+            },
+            function ($ex) use ($that, $conn) {
+                $that->emit('error', array($ex));
```

Comment:
Note, that promises can be rejected with a non-exception (there is a pending PR to change that: reactphp/promise#93). The docs suggest a typehint against `Exception` for the `error` event listeners.

```php
$http->on('error', function (Exception $e) {
});
```

Maybe check against `\Exception`/ `\Throwable` here an turn non-exceptions into a `\Exception`?